### ISSUE TITLE
CPU torch.exponential_ function may generate 0 which can cause downstream NaN

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#22557]
## 🐛 Bug

torch.exponential_ on the CPU has a small (2^-53) chance of generating 0.

I don't think we document whether torch.exponential_ may return 0, but both the [NumPy](https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.exponential.html) and [TF](https://www.tensorflow.org/api_docs/python/tf/distributions/Exponential) documentation pages don't allow for zero. This is convenient for downstream uses, such as Gumbel-softmax.

The CPU exponential generation function computes `-log(1-uniform())` (always in double-precision). The result is then rounded to the desired data type (e.g. float).

https://github.com/pytorch/pytorch/blob/042da2171e95ae5cf39452ae17a5000a5a24eb85/aten/src/ATen/core/DistributionsHelper.h#L209-L212

If `sample` is `0`, then `log(1.0 - sample)` is also `0`. Note that `sample=1` is also not allowed.

The CUDA generator doesn't have this issue because it shifts the probability mass around zero to log(1-eps).

Related to https://github.com/pytorch/pytorch/issues/22442.

## To Reproduce

I haven't reproduced this behavior. It would be nice to have specific bitstreams for RNG testing that always return specific sequences (such as 0) to test for this sort of behavior.
